src/torrent.cpp

```diff
@@ -5334,30 +5304,29 @@ namespace libtorrent
 	{
 		TORRENT_ASSERT(p != nullptr);
 		TORRENT_ASSERT(is_single_thread());
+		TORRENT_ASSERT(std::count(m_peers_to_disconnect.begin()
+			, m_peers_to_disconnect.end(), p) == 0);
 
-		auto const i = sorted_find(m_connections, p);
-		if (i == m_connections.end())
-		{
-			TORRENT_ASSERT_FAIL();
-			return;
-		}
+		std::weak_ptr<torrent> weak_t = shared_from_this();
+		m_peers_to_disconnect.push_back(p);
+		m_deferred_disconnect.post(m_ses.get_io_service(), [=]() {
```

Comment:
I would expect the ``post()`` function to be inlined, and most work of creating the handler could probably be moved by the optimizer. The one thing I don't think could be done conditionally though is the creation of the ``shared_ptr`` and ``subsequently a ``weak_ptr``